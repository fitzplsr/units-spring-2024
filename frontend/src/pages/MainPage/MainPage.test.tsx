import { fireEvent, render } from '@testing-library/react';
import { MainPage } from './MainPage';
import React from 'react';
import { Category, Product } from '../../types';
import { useCurrentTime, useProducts } from '../../hooks';
import { applyCategories, updateCategories } from '../../utils';

const products: Product[] = [
    {
        id: 1,
        name: 'IPhone 14 Pro',
        description: 'Latest iphone, buy it now',
        price: 999,
        priceSymbol: '$',
        category: 'Электроника',
        imgUrl: '/iphone.png',
    },
];

const emptyCategories: Category[] = [];

jest.mock('../../utils/getPrice', () => ({
    __esModule: true,
    getPrice: jest.fn(() => ''),
}));

jest.mock('../../utils/applyCategories', () => ({
    __esModule: true,
    applyCategories: jest.fn(() => {
        return products;
    }),
}));

jest.mock('../../utils/updateCategories', () => ({
    __esModule: true,
    updateCategories: jest.fn(() => {
        return emptyCategories;
    }),
}));

jest.mock('../../hooks/useProducts', () => ({
    __esModule: true,
    useProducts: jest.fn(() => {
        return products;
    }),
}));

jest.mock('../../hooks/useCurrentTime');

afterEach(jest.clearAllMocks);
describe('Categories test', () => {
    it('should render correctly', () => {
        const rendered = render(<MainPage />);
        expect(rendered.asFragment()).toMatchSnapshot();
    });

    it('should call callback when page render', () => {
        render(<MainPage />);
        expect(useProducts).toHaveBeenCalledTimes(1);
        expect(useCurrentTime).toHaveBeenCalledTimes(1);
        expect(applyCategories).toHaveBeenCalledWith(products, emptyCategories);
        expect(updateCategories).toHaveBeenCalledTimes(0);
    });

    it('should update categories state after click', () => {
        const rendered = render(<MainPage />);
        fireEvent.click(rendered.getByText('Одежда'));
        expect(updateCategories).toHaveBeenCalledWith(
            emptyCategories,
            'Одежда'
        );
    });
});
