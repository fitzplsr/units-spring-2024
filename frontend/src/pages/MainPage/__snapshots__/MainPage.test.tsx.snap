// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories test should render correctly 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3 />
    <div
      class="main-page__parameters"
    >
      <div
        class="categories"
      >
        <div
          class="categories__badge"
        >
          Одежда
        </div>
        <div
          class="categories__badge"
        >
          Для дома
        </div>
        <div
          class="categories__badge"
        >
          Электроника
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          IPhone 14 Pro
        </h2>
        <div
          class="product-card__description"
        >
          Latest iphone, buy it now
        </div>
        <h1
          class="product-card__price"
        />
        <div
          class="product-card__category"
        >
          Электроника
        </div>
      </div>
      <img
        alt="IPhone 14 Pro"
        class="product-card__image"
        src="/iphone.png"
      />
    </div>
  </div>
</DocumentFragment>
`;
